addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js

```diff
@@ -90,6 +90,150 @@ class EnvironmentScService extends Service {
     return this.augmentWithConnectionInfo(requestContext, envs);
   }
 
+  async pollAndSyncWsStatus(requestContext) {
+    const [indexesService, awsAccountsService] = await this.service(['indexesService', 'awsAccountsService']);
+    let envs = await this._scanner()
+      .limit(10000)
+      .scan();
+    envs = _.filter(
+      envs,
+      env => _.includes(['COMPLETED', 'STARTING', 'STOPPING', 'TERMINATING'], env.status) && env.inWorkflow !== 'true',
+    );
+    const indexes = await indexesService.list(requestContext, { fields: ['id', 'awsAccountId'] });
+    const indexesGroups = _.groupBy(indexes, index => index.awsAccountId);
+    const envGroups = _.groupBy(envs, env => env.indexId);
+    const accounts = await awsAccountsService.list(requestContext);
+    const pollAndSyncPromises = accounts.map(account =>
+      this.pollAndSyncWsStatusForAccount(requestContext, account, indexesGroups, envGroups),
+    );
+    return Promise.all(pollAndSyncPromises);
+  }
+
+  async pollAndSyncWsStatusForAccount(requestContext, account, indexesGroups, envGroups) {
+    const { roleArn, externalId, id, accountId } = account;
+    const { ec2Instances, sagemakerInstances } = this.getInstanceToCheck(_.get(indexesGroups, id), envGroups);
+    let ec2Updated = {};
+    let sagemakerUpdated = {};
+    if (!_.isEmpty(ec2Instances)) {
+      ec2Updated = await this.pollAndSyncEc2Status(roleArn, externalId, ec2Instances, requestContext);
+    }
+    if (!_.isEmpty(sagemakerInstances)) {
+      sagemakerUpdated = await this.pollAndSyncSageMakerStatus(roleArn, externalId, sagemakerInstances, requestContext);
+    }
+    return { accountId, ec2Updated, sagemakerUpdated };
+  }
+
+  async pollAndSyncEc2Status(roleArn, externalId, ec2Instances, requestContext) {
+    const EC2StatusMap = {
+      'running': 'COMPLETED',
+      'pending': 'STARTING',
+      'stopping': 'STOPPING',
+      'stopped': 'STOPPED',
+      'shutting-down': 'TERMINATING',
+      'terminated': 'TERMINATED',
+    };
+    const ec2RealtimeStatus = await this.pollEc2RealtimeStatus(roleArn, externalId, ec2Instances);
+    const ec2Updated = {};
+    _.forEach(ec2Instances, async (existingEnvRecord, key) => {
+      const expectedDDBStatus = EC2StatusMap[ec2RealtimeStatus[key]];
+      if (existingEnvRecord.status !== expectedDDBStatus) {
+        ec2Updated[key] = {
+          ddbID: existingEnvRecord.id,
+          currentStatus: expectedDDBStatus,
+          staleStatus: existingEnvRecord.status,
+        };
+        const newEnvironment = {
+          id: existingEnvRecord.id,
+          rev: existingEnvRecord.rev || 0,
```

Comment:
We are updating here but the revision stays same. Looks a little odd to me. Ideally this should be a conditional update where we fail the update in case the revision has changed. Perhaps add a TODO?

Reply:
The DB service is in charge of bumping up the rev number. When it's called with the same rev number the current record has, it bump up the rev number by one and update, when it's called with a rev number less than the current record has, it throws error.